src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionResolveHandler.cs

```diff
@@ -0,0 +1,279 @@
+            }
+
+            var document = SolutionProvider.GetDocument(data.CodeActionParams.TextDocument, clientName);
+            var codeActions = await CodeActionsHandler.GetCodeActionsAsync(
+                document,
+                _codeFixService,
+                _codeRefactoringService,
+                data.CodeActionParams.Range,
+                cancellationToken).ConfigureAwait(false);
+
+            if (codeActions == null || !codeActions.Any())
+            {
+                return codeAction;
+            }
+
+            var codeActionToResolve = GetCodeActionToResolve(data.DistinctTitle, codeActions);
+
+            // We didn't find a matching action, so just return the action without an edit or command.
+            if (codeActionToResolve == null)
+            {
+                return codeAction;
+            }
+
+            var operations = await codeActionToResolve.GetOperationsAsync(cancellationToken).ConfigureAwait(false);
+            if (operations.IsEmpty)
+            {
+                return codeAction;
+            }
+
+            // TO-DO:
+            // 1) We currently must execute code actions which add new documents on the server as commands,
+            // since there is no LSP support for adding documents yet. In the future, we should move these actions
+            // to execute on the client.
+            // 2) There is also a bug (same tracking item) where code actions that edit documents other than the
+            // one where the code action was invoked from do not work. We must temporarily execute these as commands
+            // as well.
+            // https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1147293/
+            var runAsCommand = false;
+
+            var applyChangesOperations = operations.Where(operation => operation is ApplyChangesOperation);
+            if (applyChangesOperations.Any())
+            {
+                var workspaceEdits = await ComputeWorkspaceEdits(applyChangesOperations, document!, cancellationToken).ConfigureAwait(false);
+                if (workspaceEdits.Any())
+                {
+                    codeAction.Edit = new LSP.WorkspaceEdit { DocumentChanges = workspaceEdits };
+                }
+                else
+                {
+                    // The workspace edit is something we don't currently support, like adding a new document.
+                    runAsCommand = true;
+                }
+            }
+
+            // Set up to run as command on the server instead of using WorkspaceEdits.
+            var commandOperations = operations.All(operation => !(operation is ApplyChangesOperation));
+            if (commandOperations || runAsCommand)
+            {
+                codeAction.Command = SetCommand(codeAction, data);
+            }
+
+            return codeAction;
+
+            // Local functions
+            static async Task<TextDocumentEdit[]> ComputeWorkspaceEdits(
```

Comment:
I think it would be cleaner for this to be something like TryComputeWorkspaceEdits that returns false when unable to compute them and a command has to be used instead.

Reply:
Can that be done when the function is async? If we return a bool, then we would need to have an out parameter for the workspace edits, but I don't think we can use out parameters in async methods. I guess we could also return a tuple of (bool, WorkspaceEdits) but that seems a bit messy.